tests/libvmi/vmi.go

```diff
@@ -33,8 +33,8 @@ type Option func(vmi *kvirtv1.VirtualMachineInstance)
 
 // New instantiates a new VMI configuration,
 // building its properties based on the specified With* options.
-func New(namespace, name string, opts ...Option) *kvirtv1.VirtualMachineInstance {
-	vmi := baseVmi(namespace, name)
+func New(name string, opts ...Option) *kvirtv1.VirtualMachineInstance {
```

Comment:
Creating a VMI requires a namespace explicitly. How can a caller to this differently?

The problem is with the original default which is no longer valid (in parallel run), not with the definition on the VMI namespace.

Reply:
It is contraproductive if it is set in the yaml. Especially if called like this: `return New(NamespaceTestDefault, RandName(DefaultVmiName), opts...)` inside `NewFedora(opts ...Option)`.

If I would create a VM like this:

```
vm := NewFedora()
client.VirtualMachine(tests.NamespaceTestAlternative).Create(vm)
```

the create call would fail, because the namespace in the yaml is pointing to `NamespaceTestDefault` which is different to `NamespaceTestAlternative` and I would now have to do this to make it work:

```
vm := NewFedora()
vm.Spec.Template.Metadata.Namespace = tests.NamespaceTestAlternative
client.VirtualMachine(tests.NamespaceTestAlternative).Create(vm)
```

Reply:
I am saying that the problem is in the `NewFedora` factory and not below it.
The factory should get the namespace as input explicitly and it is up to the caller to pass in the correct thing.

The current state took a shortcut as the usage did not require anything except the default.
If now it is not good enough (I guess with the computed namespace name in this case), the caller should set it explicitly.

Reply:
> I am saying that the problem is in the `NewFedora` factory and not below it.
> The factory should get the namespace as input explicitly and it is up to the caller to pass in the correct thing.
> 
> The current state took a shortcut as the usage did not require anything except the default.
> If now it is not good enough (I guess with the computed namespace name in this case), the caller should set it explicitly.
> 
> EDIT: What are the implications of not setting the namespace in the spec?

That is the whole point. You are not supposed to set the namespace inside the yaml.

Reply:
> That is the whole point. You are not supposed to set the namespace inside the yaml in this scenario. Only if you want to specify that a specific namespace needs to be used for creation.

- The VMI `New` should allow a caller to specify a specific namespace unless it is wrong to do it in general. The factory may hide it and user an empty string while others may decide to leave it up to the caller to be explicit.
- When we call `client.VirtualMachine("mynamespace").Create(vm)`, does the ns specified as the argument of `VirtualMachine` is going to overwrite what is specified in the vm spec? I do not understand how this is supposed to work. Obviously, if that argument value is enough, there is no point to set anything in the spec. Could you please clarify?

Reply:
If you set a namespace in the spec, the create calls will check if this namespace is the same which you used in the client. If they do not match the `Create` will fail. Normally one does not set the namespace in the metadata. It will be auto-filled by the cluster.

Reply:
Thank you for the clarification.